src/plugins/shutdown_button/ShutdownButton.cc

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include "ShutdownButton.hh"
+
+#include <csignal>
+#include <string>
+
+#include <ignition/common/Console.hh>
+#include <ignition/common/StringUtils.hh>
+#include <ignition/common/Time.hh>
+#include <ignition/plugin/Register.hh>
+
+#include "ignition/gui/Helpers.hh"
+
+namespace ignition
+{
+namespace gui
+{
+namespace plugins
+{
+  class ShutdownButtonPrivate
+  {
+    /// \brief Service to send server control requests
+    public: std::string controlService;
+
+    /// \brief Communication node
+    public: ignition::transport::Node node;
+  };
+}
+}
+}
+
+using namespace ignition;
+using namespace gui;
+using namespace plugins;
+
+/////////////////////////////////////////////////
+ShutdownButton::ShutdownButton()
+  : Plugin(), dataPtr(new ShutdownButtonPrivate)
+{
+}
+
+/////////////////////////////////////////////////
+ShutdownButton::~ShutdownButton()
+{
+}
+
+/////////////////////////////////////////////////
+void ShutdownButton::LoadConfig(const tinyxml2::XMLElement *_pluginElem)
+{
+  // Default name in case user didn't define one
+  if (this->title.empty())
+    this->title = "Shutdown";
+
+  // Create elements from configuration
+  if (!_pluginElem)
+  {
+    ignerr << "Null plugin element." << std::endl;
+    return;
+  }
+
+  // For server control requests
+  this->dataPtr->controlService = "/server_control";
+  auto serviceElem = _pluginElem->FirstChildElement("service");
+  if (nullptr != serviceElem && nullptr != serviceElem->GetText())
+    this->dataPtr->controlService = serviceElem->GetText();
+
+  this->dataPtr->controlService = transport::TopicUtils::AsValidTopic(
+      this->dataPtr->controlService);
+
+  if (this->dataPtr->controlService.empty())
+  {
+    ignerr << "Failed to create valid server service" << std::endl;
+    return;
+  }
+
+  ignmsg << "Using server control service [" << this->dataPtr->controlService
+         << "]" << std::endl;
+}
+
+/////////////////////////////////////////////////
+void ShutdownButton::OnStop()
+{
+  std::function<void(const ignition::msgs::Boolean &, const bool)> cb =
+      [](const ignition::msgs::Boolean &/*_rep*/, const bool /*_result*/)
+  {
+    ignmsg << "Request to shutdown the simulation server sent. Stopping client "
+           << "now." << std::endl;
+    std::raise(SIGTERM);
```

Comment:
Can we call `close()` on the window instead? `SIGTERM` feels a bit extreme if we can avoid it.

Reply:
I no longer use signals for this.